### PR TITLE
ROS#165 - HAL PIO driver

### PR DESCRIPTION
**Pull request by SteelPh0enix**
Implementation of HAL PIO driver

### CODE REVIEW

File: arch/cortex-m/samv71-hal/src/pio/output_pin.rs [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+    }
+}
+
+/// PIO-controlled pin's implementation for pin in output mode.
+///
+/// This mode allows the pin to:
+/// * Set it's output state
+/// * Configure it in open-drain or push-pull mode
+/// * Configure
+impl Pin<OutputMode> {
```
Comment (Glamhoth) [inline arch/cortex-m/samv71-hal/src/pio/output_pin.rs:100]: maybe base `impl` shoud be be before trait `impl`s? (readabilty)

File: testbins/test-hal-pio/src/main.rs [hunk 2]
```diff
@@ -0,0 +1,234 @@
...
+    },
+    Aerugo, InitApi, SystemHardwareConfig,
+};
+use calldwell::with_rtt_out;
+use rt::entry;
+
+fn validate_pin_is_driven_high(pin: &Pin<OutputMode>) {
+    if !pin.is_set_high().unwrap() {
+        panic!("Pin is driven high, yet `is_set_high` returns `false`");
+    }
```
Comment (Glamhoth) [inline testbins/test-hal-pio/src/main.rs:26]: Those could be replaced with
`assert!(cond, "message")`

**Review by Glamhoth — CHANGES_REQUESTED** [review]

File: arch/cortex-m/samv71-hal/src/pio/output_pin.rs (hunk 1 above)
Comment (SteelPh0enix) [reply arch/cortex-m/samv71-hal/src/pio/output_pin.rs:100]: Most common functions are implemented in trait impls, this is just additional interface, so it's reasonable to keep more commonly used implementations up there.

File: testbins/test-hal-pio/src/main.rs (hunk 2 above)
Comment (SteelPh0enix) [reply testbins/test-hal-pio/src/main.rs]: fix'd

**Review by Glamhoth — APPROVED** [review]

**Pull request merged**